botocore/utils.py

```diff
@@ -919,6 +961,8 @@ def is_valid_endpoint_url(endpoint_url):
         re.IGNORECASE)
     return allowed.match(hostname)
 
+def is_valid_ipv4_or_ipv6_endpoint(endpoint_url):
```

Comment:
We've had some back and forth on this, and I think this name still seems a bit wrong. The endpoint isn't necessarily an ip address and` is_valid_endpoint_url` isn't specific to IPv4.

Maybe we should either name this something like `is_valid_uri` which is a more complete set of rules compared to `is_valid_endpoint_url`. We could also update the definition to `is_valid_endpoint_url` to be this combination, but that may break expectations for users who are currently using it.